ledger/participant-integration-api/src/main/scala/platform/store/appendonlydao/events/ContractsReader.scala

```diff
@@ -203,28 +206,36 @@ private[appendonlydao] sealed class ContractsReader(
   override def lookupContractKey(
       key: Key,
       readers: Set[Party],
+      ledgerEndSequentialId: Option[EventSequentialId] = None,
   )(implicit loggingContext: LoggingContext): Future[Option[ContractId]] =
     Timed.future(
       metrics.daml.index.db.lookupKey,
       dispatcher.executeSql(metrics.daml.index.db.lookupContractByKeyDbMetrics) {
         implicit connection =>
-          lookupContractKeyQuery(readers, key).as(contractId("contract_id").singleOpt)
+          lookupContractKeyQuery(readers, key, ledgerEndSequentialId).as(
+            contractId("contract_id").singleOpt
+          )
       },
     )
 
   private def lookupActiveContractAndLoadArgumentQuery(
       contractId: ContractId,
       readers: Set[Party],
+      // TODO Remove SQL references to `parameters` table when removing the `ledgerEndSequentialId` parameter optionality
+      ledgerEndSequentialId: Option[EventSequentialId],
   ) = {
     val tree_event_witnessesWhere =
       sqlFunctions.arrayIntersectionWhereClause("tree_event_witnesses", readers)
+
+    val ledgerEndLimit =
+      ledgerEndSequentialId.map(_.toString).getOrElse("parameters.ledger_end_sequential_id")
     SQL"""
   WITH archival_event AS (
          SELECT participant_events.*
            FROM participant_events, parameters
```

Comment:
You are unconditionally joining with the parameters table. Maybe something like this?
```suggestion
           FROM participant_events#${ledgerEndSequentialId.map(_ => "").getOrElse(", parameters")}
```

Reply:
I was thinking it has no impact if the table is not actually queried. Fine to add it as well.